File: 2022/Day8/python/Day8_golf.py
# Part 1 and Part 2

# Zheng Lin Lei

grid = [list(map(int, l)) for l in open('input').read().split('\n')]
height, width = len(grid), len(grid[0])
visible = [[False for _ in range(width)] for _ in range(height)]
best_scenic = 0

for x0 in range(width):
    for y0 in range(height):
        ss, h = 1, grid[y0][x0]
        for dx, dy in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
            x1, y1, s1 = x0+dx, y0+dy, 1
            while 0 <= x1 < width and 0 <= y1 < height and grid[y1][x1] < h:
                x1, y1, s1 = x1 + dx, y1 + dy, s1 + 1
            if not (0 <= x1 < width and 0 <= y1 < height):
                visible[y0][x0] = True
                s1 -= 1
            ss *= s1
        best_scenic = max(best_scenic, ss)

print(sum(map(sum, visible)))
print(best_scenic)

File: 2022/Day8/javascript/Day8_golf.js
// Read File ../input.txt

// Part 1 and Part 2


// ZhengLinLei

const fs = require('fs');

// Read File Async
fs.readFile('../input.txt', 'utf8', (err, text) =>  {
    z=()=>text.split(/\n/).map(r=>[...r]);
    w=(m)=>m[0].map((_,i)=>m.map(r=>r[98-i]));s=z();f=z();
    for(i=0;i++<4;){
        for(y in f)for(x in f[y]) if(f[y].slice(0,x).every(v=>+v<+f[y][x])) s[y][x]='x';
        f=w(f);s=w(s);
    }console.log((""+s).split('x').length-1);
    f=text.split(/\n/).map(r=>[...r]);
    w=(m)=>m[0].map((_,i)=>m.map(r=>r[98-i]));s=f.map(r=>r.map(c=>1));
    for(i=0;i++<4;){
        for(y in f)for(x in f[y])s[y][x]*=(1+f[y].slice(+x+1).findIndex(v=>v>=f[y][x]))||98-x
        f=w(f);s=w(s);
    }console.log(Math.max(...(""+s).split(',')));
});